cli/src/cli_util.rs

```diff
@@ -1148,17 +1149,19 @@ impl WorkspaceCommandHelper {
         Ok(revset_expression.evaluate(self.repo().as_ref())?)
     }
 
-    pub(crate) fn revset_parse_context(&self) -> RevsetParseContext {
+    pub(crate) fn revset_parse_context(&self) -> Result<RevsetParseContext, CommandError> {
         let workspace_context = RevsetWorkspaceContext {
             cwd: &self.cwd,
             workspace_id: self.workspace_id(),
             workspace_root: self.workspace.workspace_root(),
         };
-        RevsetParseContext {
+        let context = RevsetParseContext {
             aliases_map: &self.revset_aliases_map,
             user_email: self.settings.user_email(),
+            immutable_heads: Some(self.settings.immutable_heads_revset()),
```

Comment:
I actually did that at first but I didn't like the idea of parsing it if it's not used in the main expression. What are the advantages of passing in a parsed expression?

Reply:
I just feel it would be better to not recurse into parser from arbitrary function.

Since `immutable()` is basically an alias, it could be defined as an alias. It would provide better error message. The downside is that we cannot enforce that the immutable set includes `root()`. I guess that's the original idea you've rejected.

Reply:
> I just feel it would be better to not recurse into parser from arbitrary function.

Okay, I'll switch back to passing in a `RevsetExpression` in the context.

> The downside is that we cannot enforce that the immutable set includes `root()`.

Yes, I wanted to make sure that `immutable()` always includes `root()` and all ancestors of the configured immutable heads.